File: swift/tree234.swift
// Generic array-backed 2-3-4 tree with monoidal measures.
//
// Type parameters:
//   K - key/element type (stored in sorted order)
//   M - measure type (monoidal annotation on subtrees)
//
// Traits protocol provides:
//   measure(K) -> M         - measure a single element
//   combine(M, M) -> M     - monoidal combine
//   identity() -> M        - monoidal identity
//   compare(K, K) -> Int   - <0, 0, >0

protocol Tree234Traits {
    associatedtype K
    associatedtype M

    static func measure(_ key: K) -> M
    static func combine(_ a: M, _ b: M) -> M
    static func identity() -> M
    static func compare(_ a: K, _ b: K) -> Int
}

struct WeightResult<K> {
    var key: K
    var cumBefore: Double
    var index: Int
    var found: Bool
}

struct Tree234<Traits: Tree234Traits> {
    typealias K = Traits.K
    typealias M = Traits.M

    struct Node {
        var n: Int = 0             // number of keys: 1, 2, or 3
        var keys: (K?, K?, K?) = (nil, nil, nil)
        var children: (Int, Int, Int, Int) = (-1, -1, -1, -1)
        var measure: M = Traits.identity()

        // Key accessors using tuples (Swift doesn't allow fixed-size arrays easily)
        func key(_ i: Int) -> K {
            switch i {
            case 0: return keys.0!
            case 1: return keys.1!
            case 2: return keys.2!
            default: fatalError("key index out of range")
            }
        }

        mutating func setKey(_ i: Int, _ v: K) {
            switch i {
            case 0: keys.0 = v
            case 1: keys.1 = v
            case 2: keys.2 = v
            default: fatalError("key index out of range")
            }
        }

        func child(_ i: Int) -> Int {
            switch i {
            case 0: return children.0
            case 1: return children.1
            case 2: return children.2
            case 3: return children.3
            default: fatalError("child index out of range")
            }
        }

        mutating func setChild(_ i: Int, _ v: Int) {
            switch i {
            case 0: children.0 = v
            case 1: children.1 = v
            case 2: children.2 = v
            case 3: children.3 = v
            default: fatalError("child index out of range")
            }
        }
    }

    private var nodes: [Node] = []
    private var freeList: [Int] = []
    private var root: Int = -1
    private(set) var count: Int = 0

    init() {}

    // MARK: - Node allocation

    private mutating func allocNode() -> Int {
        if let idx = freeList.popLast() {
            nodes[idx] = Node()
            return idx
        }
        let idx = nodes.count
        nodes.append(Node())
        return idx
    }

    private mutating func freeNode(_ idx: Int) {
        freeList.append(idx)
    }

    private func isLeaf(_ idx: Int) -> Bool {
        return nodes[idx].children.0 == -1
    }

    private func is4Node(_ idx: Int) -> Bool {
        return nodes[idx].n == 3
    }

    // MARK: - Measure recomputation

    private mutating func recomputeMeasure(_ idx: Int) {
        let nd = nodes[idx]
        var m = Traits.identity()
        for i in 0...nd.n {
            let c = nd.child(i)
            if c != -1 {
                m = Traits.combine(m, nodes[c].measure)
            }
            if i < nd.n {
                m = Traits.combine(m, Traits.measure(nd.key(i)))
            }
        }
        nodes[idx].measure = m
    }

    // MARK: - Split 4-node child

    private mutating func splitChild(_ parentIdx: Int, _ childPos: Int) {
        let childIdx = nodes[parentIdx].child(childPos)

        // Save child data before allocNode may grow the array
        let k0 = nodes[childIdx].key(0)
        let k1 = nodes[childIdx].key(1)
        let k2 = nodes[childIdx].key(2)
        let c0 = nodes[childIdx].child(0)
        let c1 = nodes[childIdx].child(1)
        let c2 = nodes[childIdx].child(2)
        let c3 = nodes[childIdx].child(3)

        // Create right node with k2, c2, c3
        let rightIdx = allocNode()
        nodes[rightIdx].n = 1
        nodes[rightIdx].setKey(0, k2)
        nodes[rightIdx].setChild(0, c2)
        nodes[rightIdx].setChild(1, c3)

        // Shrink child (left) to k0, c0, c1
        nodes[childIdx].n = 1
        nodes[childIdx].setKey(0, k0)
        nodes[childIdx].setChild(0, c0)
        nodes[childIdx].setChild(1, c1)
        nodes[childIdx].setChild(2, -1)
        nodes[childIdx].setChild(3, -1)

        recomputeMeasure(childIdx)
        recomputeMeasure(rightIdx)

        // Insert mid key (k1) into parent at childPos
        let parentN = nodes[parentIdx].n
        // Shift keys and children right
        var i = parentN
        while i > childPos {
            nodes[parentIdx].setKey(i, nodes[parentIdx].key(i - 1))
            nodes[parentIdx].setChild(i + 1, nodes[parentIdx].child(i))
            i -= 1
        }
        nodes[parentIdx].setKey(childPos, k1)
        nodes[parentIdx].setChild(childPos + 1, rightIdx)
        nodes[parentIdx].n += 1

        recomputeMeasure(parentIdx)
    }

    // MARK: - Insert

    private mutating func insertNonFull(_ idx: Int, _ key: K) {
        if isLeaf(idx) {
            var pos = nodes[idx].n
            while pos > 0 && Traits.compare(key, nodes[idx].key(pos - 1)) < 0 {
                nodes[idx].setKey(pos, nodes[idx].key(pos - 1))
                pos -= 1
            }
            nodes[idx].setKey(pos, key)
            nodes[idx].n += 1
            recomputeMeasure(idx)
            return
        }

        var pos = 0
        while pos < nodes[idx].n && Traits.compare(key, nodes[idx].key(pos)) >= 0 {
            pos += 1
        }

        if is4Node(nodes[idx].child(pos)) {
            splitChild(idx, pos)
            if Traits.compare(key, nodes[idx].key(pos)) >= 0 {
                pos += 1
            }
        }

        insertNonFull(nodes[idx].child(pos), key)
        recomputeMeasure(idx)
    }

    mutating func insert(_ key: K) {
        if root == -1 {
            root = allocNode()
            nodes[root].n = 1
            nodes[root].setKey(0, key)
            recomputeMeasure(root)
            count += 1
            return
        }

        if is4Node(root) {
            let oldRoot = root
            root = allocNode()
            nodes[root].setChild(0, oldRoot)
            splitChild(root, 0)
        }

        insertNonFull(root, key)
        count += 1
    }

    // MARK: - Clear

    mutating func clear() {
        nodes.removeAll()
        freeList.removeAll()
        root = -1
        count = 0
    }

    // MARK: - Root measure

    func rootMeasure() -> M {
        if root == -1 { return Traits.identity() }
        return nodes[root].measure
    }

    // MARK: - In-order traversal

    private func forEachImpl(_ idx: Int, _ body: (K) -> Void) {
        if idx == -1 { return }
        let nd = nodes[idx]
        for i in 0...nd.n {
            let c = nd.child(i)
            if c != -1 {
                forEachImpl(c, body)
            }
            if i < nd.n {
                body(nd.key(i))
            }
        }
    }

    func forEach(_ body: (K) -> Void) {
        forEachImpl(root, body)
    }

    // MARK: - Collect

    func collect() -> [K] {
        var result: [K] = []
        result.reserveCapacity(count)
        forEach { result.append($0) }
        return result
    }

    // MARK: - Subtree count

    private func subtreeCount(_ idx: Int) -> Int {
        if idx == -1 { return 0 }
        let nd = nodes[idx]
        var c = nd.n
        for i in 0...nd.n {
            let ch = nd.child(i)
            if ch != -1 {
                c += subtreeCount(ch)
            }
        }
        return c
    }

    // MARK: - Find by weight

    private func findByWeightImpl(_ idx: Int, _ target: Double, _ cum: Double,
                                   _ globalIdx: Int,
                                   _ weightOf: (M) -> Double) -> WeightResult<K>? {
        if idx == -1 { return nil }

        let nd = nodes[idx]
        var runningCum = cum
        var runningIdx = globalIdx

        for i in 0...nd.n {
            let c = nd.child(i)
            if c != -1 {
                let childWeight = weightOf(nodes[c].measure)
                if runningCum + childWeight >= target {
                    return findByWeightImpl(c, target, runningCum, runningIdx, weightOf)
                }
                runningCum += childWeight
                runningIdx += subtreeCount(c)
            }

            if i < nd.n {
                let keyWeight = weightOf(Traits.measure(nd.key(i)))
                if runningCum + keyWeight >= target {
                    return WeightResult(key: nd.key(i), cumBefore: runningCum,
                                        index: runningIdx, found: true)
                }
                runningCum += keyWeight
                runningIdx += 1
            }
        }

        return nil
    }

    func findByWeight(_ target: Double, _ weightOf: (M) -> Double) -> WeightResult<K>? {
        if root == -1 { return nil }
        return findByWeightImpl(root, target, 0.0, 0, weightOf)
    }

    // MARK: - Build from sorted

    private mutating func buildRecursive(_ sorted: [K], _ lo: Int, _ hi: Int) -> Int {
        let n = hi - lo
        if n <= 0 { return -1 }

        if n <= 3 {
            let idx = allocNode()
            nodes[idx].n = n
            for i in 0..<n {
                nodes[idx].setKey(i, sorted[lo + i])
            }
            recomputeMeasure(idx)
            return idx
        }

        if n <= 7 {
            let mid = lo + n / 2
            let left = buildRecursive(sorted, lo, mid)
            let right = buildRecursive(sorted, mid + 1, hi)
            let idx = allocNode()
            nodes[idx].n = 1
            nodes[idx].setKey(0, sorted[mid])
            nodes[idx].setChild(0, left)
            nodes[idx].setChild(1, right)
            recomputeMeasure(idx)
            return idx
        }

        // Use 3-node for larger ranges
        let third = n / 3
        let m1 = lo + third
        let m2 = lo + 2 * third + 1
        let c0 = buildRecursive(sorted, lo, m1)
        let c1 = buildRecursive(sorted, m1 + 1, m2)
        let c2 = buildRecursive(sorted, m2 + 1, hi)
        let idx = allocNode()
        nodes[idx].n = 2
        nodes[idx].setKey(0, sorted[m1])
        nodes[idx].setKey(1, sorted[m2])
        nodes[idx].setChild(0, c0)
        nodes[idx].setChild(1, c1)
        nodes[idx].setChild(2, c2)
        recomputeMeasure(idx)
        return idx
    }

    mutating func buildFromSorted(_ sorted: [K]) {
        clear()
        if sorted.isEmpty { return }
        count = sorted.count
        root = buildRecursive(sorted, 0, sorted.count)
    }

    // MARK: - Size

    var size: Int { return count }
}

File: csharp/Tree234.cs
// Generic array-backed 2-3-4 tree with monoidal measures.
//
// Type parameters:
//   K      - key/element type (stored in sorted order)
//   M      - measure type (monoidal annotation on subtrees)
//
// ITree234Traits<K, M> provides:
//   M Measure(K)          - measure a single element
//   M Combine(M, M)       - monoidal combine
//   M Identity()           - monoidal identity
//   int Compare(K, K)      - <0, 0, >0

using System;
using System.Collections.Generic;

namespace TDigestLib
{
    public interface ITree234Traits<K, M>
    {
        M Measure(K key);
        M Combine(M a, M b);
        M Identity();
        int Compare(K a, K b);
    }

    public struct WeightResult<K>
    {
        public K Key;
        public double CumBefore;
        public int Index;
        public bool Found;
    }

    public class Tree234<K, M, TTraits> where TTraits : struct, ITree234Traits<K, M>
    {
        private struct Node
        {
            public int N;               // number of keys: 1, 2, or 3
            public K Key0, Key1, Key2;
            public int Child0, Child1, Child2, Child3;
            public M Measure;

            public K GetKey(int i)
            {
                switch (i) { case 0: return Key0; case 1: return Key1; default: return Key2; }
            }

            public void SetKey(int i, K val)
            {
                switch (i) { case 0: Key0 = val; break; case 1: Key1 = val; break; default: Key2 = val; break; }
            }

            public int GetChild(int i)
            {
                switch (i) { case 0: return Child0; case 1: return Child1; case 2: return Child2; default: return Child3; }
            }

            public void SetChild(int i, int val)
            {
                switch (i) { case 0: Child0 = val; break; case 1: Child1 = val; break; case 2: Child2 = val; break; default: Child3 = val; break; }
            }
        }

        private TTraits _traits;
        private List<Node> _nodes;
        private List<int> _freeList;
        private int _root;
        private int _count;

        public Tree234()
        {
            _traits = default;
            _nodes = new List<Node>();
            _freeList = new List<int>();
            _root = -1;
            _count = 0;
        }

        private Node NewNode()
        {
            var nd = new Node();
            nd.N = 0;
            nd.Child0 = -1; nd.Child1 = -1; nd.Child2 = -1; nd.Child3 = -1;
            nd.Measure = _traits.Identity();
            return nd;
        }

        private int AllocNode()
        {
            int idx;
            if (_freeList.Count > 0)
            {
                idx = _freeList[_freeList.Count - 1];
                _freeList.RemoveAt(_freeList.Count - 1);
                _nodes[idx] = NewNode();
            }
            else
            {
                idx = _nodes.Count;
                _nodes.Add(NewNode());
            }
            return idx;
        }

        private bool IsLeaf(int idx) => _nodes[idx].Child0 == -1;
        private bool Is4Node(int idx) => _nodes[idx].N == 3;

        private void RecomputeMeasure(int idx)
        {
            var nd = _nodes[idx];
            M m = _traits.Identity();
            for (int i = 0; i <= nd.N; i++)
            {
                int child = nd.GetChild(i);
                if (child != -1)
                    m = _traits.Combine(m, _nodes[child].Measure);
                if (i < nd.N)
                    m = _traits.Combine(m, _traits.Measure(nd.GetKey(i)));
            }
            nd.Measure = m;
            _nodes[idx] = nd;
        }

        private void SplitChild(int parentIdx, int childPos)
        {
            var parent = _nodes[parentIdx];
            int childIdx = parent.GetChild(childPos);
            var child = _nodes[childIdx];

            K k0 = child.Key0, k1 = child.Key1, k2 = child.Key2;
            int c0 = child.Child0, c1 = child.Child1, c2 = child.Child2, c3 = child.Child3;

            int rightIdx = AllocNode();
            // After AllocNode, re-read parent/child by index
            var right = _nodes[rightIdx];
            right.N = 1;
            right.Key0 = k2;
            right.Child0 = c2;
            right.Child1 = c3;
            _nodes[rightIdx] = right;

            // Shrink child (left) to k0, c0, c1
            child = _nodes[childIdx];
            child.N = 1;
            child.Key0 = k0;
            child.Child0 = c0;
            child.Child1 = c1;
            child.Child2 = -1;
            child.Child3 = -1;
            _nodes[childIdx] = child;

            RecomputeMeasure(childIdx);
            RecomputeMeasure(rightIdx);

            // Insert k1 into parent at childPos
            parent = _nodes[parentIdx];
            for (int i = parent.N; i > childPos; i--)
            {
                parent.SetKey(i, parent.GetKey(i - 1));
                parent.SetChild(i + 1, parent.GetChild(i));
            }
            parent.SetKey(childPos, k1);
            parent.SetChild(childPos + 1, rightIdx);
            parent.N++;
            _nodes[parentIdx] = parent;

            RecomputeMeasure(parentIdx);
        }

        private void InsertNonFull(int idx, K key)
        {
            var nd = _nodes[idx];
            if (IsLeaf(idx))
            {
                int pos = nd.N;
                while (pos > 0 && _traits.Compare(key, nd.GetKey(pos - 1)) < 0)
                {
                    nd.SetKey(pos, nd.GetKey(pos - 1));
                    pos--;
                }
                nd.SetKey(pos, key);
                nd.N++;
                _nodes[idx] = nd;
                RecomputeMeasure(idx);
                return;
            }

            int p = 0;
            while (p < nd.N && _traits.Compare(key, nd.GetKey(p)) >= 0)
                p++;

            if (Is4Node(nd.GetChild(p)))
            {
                SplitChild(idx, p);
                nd = _nodes[idx]; // re-read after split
                if (_traits.Compare(key, nd.GetKey(p)) >= 0)
                    p++;
            }

            InsertNonFull(nd.GetChild(p), key);
            RecomputeMeasure(idx);
        }

        public void Insert(K key)
        {
            if (_root == -1)
            {
                _root = AllocNode();
                var nd = _nodes[_root];
                nd.N = 1;
                nd.Key0 = key;
                _nodes[_root] = nd;
                RecomputeMeasure(_root);
                _count++;
                return;
            }

            if (Is4Node(_root))
            {
                int oldRoot = _root;
                _root = AllocNode();
                var nd = _nodes[_root];
                nd.Child0 = oldRoot;
                _nodes[_root] = nd;
                SplitChild(_root, 0);
            }

            InsertNonFull(_root, key);
            _count++;
        }

        public void Clear()
        {
            _nodes.Clear();
            _freeList.Clear();
            _root = -1;
            _count = 0;
        }

        public int Size => _count;

        public M RootMeasure
        {
            get
            {
                if (_root == -1) return _traits.Identity();
                return _nodes[_root].Measure;
            }
        }

        private void ForEachImpl(int idx, Action<K> action)
        {
            if (idx == -1) return;
            var nd = _nodes[idx];
            for (int i = 0; i <= nd.N; i++)
            {
                int child = nd.GetChild(i);
                if (child != -1)
                    ForEachImpl(child, action);
                if (i < nd.N)
                    action(nd.GetKey(i));
            }
        }

        public void ForEach(Action<K> action)
        {
            ForEachImpl(_root, action);
        }

        public void Collect(List<K> output)
        {
            output.Clear();
            ForEach(k => output.Add(k));
        }

        private int SubtreeCount(int idx)
        {
            if (idx == -1) return 0;
            var nd = _nodes[idx];
            int c = nd.N;
            for (int i = 0; i <= nd.N; i++)
            {
                int child = nd.GetChild(i);
                if (child != -1)
                    c += SubtreeCount(child);
            }
            return c;
        }

        public WeightResult<K> FindByWeight(double target, Func<M, double> weightOf)
        {
            if (_root == -1)
                return new WeightResult<K> { Found = false };
            return FindByWeightImpl(_root, target, 0.0, 0, weightOf);
        }

        private WeightResult<K> FindByWeightImpl(int idx, double target, double cum, int globalIdx, Func<M, double> weightOf)
        {
            if (idx == -1)
                return new WeightResult<K> { Found = false };

            var nd = _nodes[idx];
            double runningCum = cum;
            int runningIdx = globalIdx;

            for (int i = 0; i <= nd.N; i++)
            {
                int child = nd.GetChild(i);
                if (child != -1)
                {
                    double childWeight = weightOf(_nodes[child].Measure);
                    if (runningCum + childWeight >= target)
                        return FindByWeightImpl(child, target, runningCum, runningIdx, weightOf);
                    runningCum += childWeight;
                    runningIdx += SubtreeCount(child);
                }

                if (i < nd.N)
                {
                    double keyWeight = weightOf(_traits.Measure(nd.GetKey(i)));
                    if (runningCum + keyWeight >= target)
                        return new WeightResult<K> { Key = nd.GetKey(i), CumBefore = runningCum, Index = runningIdx, Found = true };
                    runningCum += keyWeight;
                    runningIdx++;
                }
            }

            return new WeightResult<K> { Found = false };
        }

        public void BuildFromSorted(List<K> sorted)
        {
            Clear();
            if (sorted.Count == 0) return;
            _count = sorted.Count;
            _root = BuildRecursive(sorted, 0, sorted.Count);
        }

        private int BuildRecursive(List<K> sorted, int lo, int hi)
        {
            int n = hi - lo;
            if (n <= 0) return -1;

            if (n <= 3)
            {
                int idx = AllocNode();
                var nd = _nodes[idx];
                nd.N = n;
                for (int i = 0; i < n; i++)
                    nd.SetKey(i, sorted[lo + i]);
                _nodes[idx] = nd;
                RecomputeMeasure(idx);
                return idx;
            }

            if (n <= 7)
            {
                int mid = lo + n / 2;
                int left = BuildRecursive(sorted, lo, mid);
                int right = BuildRecursive(sorted, mid + 1, hi);
                int idx = AllocNode();
                var nd = _nodes[idx];
                nd.N = 1;
                nd.Key0 = sorted[mid];
                nd.Child0 = left;
                nd.Child1 = right;
                _nodes[idx] = nd;
                RecomputeMeasure(idx);
                return idx;
            }

            // 3-node for larger ranges
            int third = n / 3;
            int m1 = lo + third;
            int m2 = lo + 2 * third + 1;
            int c0 = BuildRecursive(sorted, lo, m1);
            int c1 = BuildRecursive(sorted, m1 + 1, m2);
            int c2 = BuildRecursive(sorted, m2 + 1, hi);
            {
                int idx = AllocNode();
                var nd = _nodes[idx];
                nd.N = 2;
                nd.Key0 = sorted[m1];
                nd.Key1 = sorted[m2];
                nd.Child0 = c0;
                nd.Child1 = c1;
                nd.Child2 = c2;
                _nodes[idx] = nd;
                RecomputeMeasure(idx);
                return idx;
            }
        }
    }
}
